packages/astro/src/runtime/client/dev-overlay/plugins/audit.ts

```diff
@@ -0,0 +1,119 @@
+
+interface AuditRule {
+	title: string;
+	message: string;
+}
+
+const selectorBasedRules: (AuditRule & { selector: string })[] = [
+	{
+		title: 'Missing `alt` tag',
+		message: 'The alt attribute is important for accessibility.',
+		selector: 'img:not([alt])',
+	},
+];
+
+export default {
+	id: 'astro:audit',
+	name: 'Audit',
+	icon: icon,
+	init(canvas, eventTarget) {
+		let audits: { highlightElement: DevOverlayHighlight; auditedElement: HTMLElement }[] = [];
+
+		selectorBasedRules.forEach((rule) => {
+			document.querySelectorAll(rule.selector).forEach((el) => {
+				createAuditProblem(rule, el);
+			});
+		});
+
+		if (audits.length > 0) {
+			eventTarget.dispatchEvent(
+				new CustomEvent('plugin-notification', {
+					detail: {
+						state: true,
+					},
+				})
+			);
+		}
+
+		function createAuditProblem(rule: AuditRule, originalElement: Element) {
+			const computedStyle = window.getComputedStyle(originalElement);
+			const targetedElement = (originalElement.children[0] as HTMLElement) || originalElement;
+
+			// If the element is hidden, don't do anything
+			if (targetedElement.offsetParent === null || computedStyle.display === 'none') {
+				return;
+			}
+
+			const highlight = document.createElement('astro-overlay-highlight') as DevOverlayHighlight;
+			highlight.icon = 'warning';
+
+			const rect = originalElement.getBoundingClientRect();
+
+			// Make an highlight that is 10px bigger than the element on all sides
+			highlight.style.top = `${Math.max(rect.top + window.scrollY - 10, 0)}px`;
+			highlight.style.left = `${Math.max(rect.left + window.scrollX - 10, 0)}px`;
+			highlight.style.width = `${rect.width + 15}px`;
+			highlight.style.height = `${rect.height + 15}px`;
+
+			const tooltip = document.createElement('astro-overlay-tooltip') as DevOverlayTooltip;
+			tooltip.sections = [
+				{
+					icon: 'warning',
+					title: rule.title,
+				},
+				{
+					content: rule.message,
+				},
+				// TODO: Add a link to the file
+				// Needs https://github.com/withastro/compiler/pull/375
+				// {
+				// 	content: '/src/somewhere/component.astro',
+				// 	clickDescription: 'Click to go to file',
+				// 	clickAction() {},
+				// },
+			];
+
+			highlight.shadowRoot.appendChild(tooltip);
+
+			highlight.addEventListener('mouseover', () => {
+				tooltip.dataset.show = 'true';
```

Comment:
I'm sure @Yan-Thomas has thoughts on this but I know tooltips have some accessibility gotchas. It's usually wise to implement a short delay between `mouseover` and toggling the visibility.

Reply:
I'm not sure a delay makes sense here because the highlight by itself shows nothing and you only have the feature enabled when you're planning to use it, so you can't hover things by accident really, right.

But yes, I'm definitely happy to consider accessibility here! I purposely didn't go on it too hard yet, apart from the keyboard navigation on the actual bar itself as I was focusing on features, but it'll be a major part of this work, especially as more components are implemented for people to use.

Reply:
That makes sense, definitely something to iterate on! I admittedly need to try this out locally to see how the instant toggle feels, but I'm sure it's fine for now.